posts/intro-to-mathport.md

```diff
@@ -0,0 +1,131 @@
+    That will hopefully come later, but we need to fix many mathport issues first.
+  * Instead, it is fine to make changes *that you think mathport should be doing already* and committing these on a branch,
+    so that you can link to diffs when opening mathport issues.
+  * For now, don't worry too much about the state of proofs.
+    We'd like to get to a state where the vast majority of *statements* are correctly translated as soon as possible.
+  * There are still many alignment problems between Lean 3 / mathlib3 declarations, and Lean 4 / mathlib4 declarations.
+    Sometimes these can be fixed by adding `#align` commands in mathlib4. Sometimes they may turn out to be mathport bugs.
+    Sometimes they will reflect deeper design problems we're going to need to talk about!
+
+
+# Background: what is mathport?
+
+`mathport`, consists of two loosely coupled components: `binport` (largely Daniel's work), and `synport` (largely Mario's work).
+
+* `binport` constructs Lean 4 `.olean` files, from Lean 3 `.olean` files.
+  It largely works, and means that you can import mathlib3 content into Lean 4 (as long as you don't expect to have source files!)
+  This is what lets us do things like:
+```
+import Mathbin
+
+#lookup3 algebraic_geometry.Scheme
+#check AlgebraicGeometry.Scheme
+```
+  Yay, Lean 4 has schemes! :-)
+  To see this file in action, you should check out a copy of the `mathlib3port` repository described below,
+  and make a new file there.
+* `synport` constructs Lean 4 `.lean` files, on a "best effort" basis.
+  (It uses both the output from binport, and Lean 3's `--ast` output to guide it.)
+  We should not expect that this will ever converge to a perfect translator.
+  Instead the hope is that it gives us something that humans can plausibly improve to a complete translation of mathlib3.
+
+To understand how mathport (mostly talking about the synport part from here on) works,
+it's important to understand that it is translating mathlib3 to Lean 4 source code, "modulo" the current content of mathlib4.
+That is, the premise is that as we progressively construct mathlib4
+(whether by translating by hand, moving content from mathport's output to mathlib4, or adding `#align` statements)
+the output from running mathport on mathlib3 will change.
+In particular, as mathport is translating each declaration,
+it checks to see if a corresponding declaration in mathlib4 already exists, and is defeq.
+If so, mathport will instead just use that declaration.
+If not, mathport will make a copy, appending a `×` to the name.
+Sometimes these misalignments are due to an unintentional non-defeq, that can be fixed in mathlib4.
+Other times, we genuinely want to change something in mathlib4
+(e.g. to use Lean 4's multiple parent structures, which are better than `old_structure_cmd`).
+As a result, we expect that some misalignments will persist throughout the mathport-assisted stage of the port,
+and only afterwards will we polish these away.
+
```

Comment:
Some readers will be very curious to learn more about how each of these are implemented. It would be great to add more detail; barring that, linking to other sources, e.g. Mario's talk at https://www.youtube.com/watch?v=fOO2fByx8tw ?

Reply:
I've added this link, and an apology. I'll ask Mario and Daniel if they would like to write more, either in this blog post or as a follow-up. I suspect neither have the time right now.